### ISSUE TITLE
Insufficient balance. Had 652814980819318111, needed 652814980819318144

### ISSUE THREAD

**Issue by alb2001** [uniswap-python/uniswap-python#323]
**Describe the bug**
Insufficient balance. Had 652814980819318111, needed 652814980819318144 when trying to swap a token back to eth

**To Reproduce**
Goerli network, UNI token on Uniswap
buy worth 0.1 of UNI token using
`self.uniswap.make_trade(ETH, self.contract.address, int(self.baseamount_Entry_var.get() * 10 ** 18), slippage=slippage)`

the amount of raw tokens I get when querying the balance is "652814980819318111", then I do
`balance = balance / 10 ** self.contract.decimals` and end up with "0.6528149808193181" (I miss the 2 last decimals here). Not sure why this is the case as this should return the same number as in etherscan, as the decimals are 18

When trying to sell them using
`self.uniswap.make_trade(self.contract.address, ETH, int(self.tokenamount_Entry_var.get() * 10 ** self.contract.decimals), slippage=slippage)`

I sometimes get, althought not always, a exception like the following:
Insufficient balance. Had 652814980819318111, needed 652814980819318144

Not sure if there's a bug here, but I end up having to take one of the last decimals on the balance box to be able to swap this token back to eth but this always leaves dust in my wallet.. Why is it needing more than I have where the number is less than max, or equal?

**Expected behavior**
Swap back and not leave any dust on my wallet

**Additional context**
Add any other context about the problem here, such as a transaction ID demonstrating the problem.

**Comment by liquid-8** [comment]
That's not a bug. Try to use decimal instead of float in `baseamount_Entry_var.get()`.

**Comment by alb2001** [comment]
Hi

I think you mean `tokenamount_Entry_var.get()` as its the one that holds the token amount I want to sell. Actually that one holds the token amount

Ok, so using Decimal has indeed solved the precision issue, now I get 0.652814980819318111 as the decimal number, which is correct, instead of 0.6528149808193181 - Missing the last 2 precision points. So that's solved.

However, when casting it back to an int for selling by doing:
`int(0.652814980819318111 * 10 ** 18)` returns 652814980819318144. This is way beyond than the original raw number 652814980819318111

How can I solve this whe npassing into an int raw number, and why is way off?

Thank you

**Comment by ErikBjare** [comment]
You won't get enough accuracy using a normal Python float to do the multiplication. Try using Decimal("0.1111111...") instead:

![IMG_20230309_131817.jpg](https://user-images.githubusercontent.com/1405370/224021123-a8af0c08-6cd8-4ac8-9a64-9796aabfd35c.jpg)

**Comment by alb2001** [comment]
Oh, so by converting the number to a string?

So far I tried the following:

`>>> amount = 0.652814980819318111`
Assigned to variable for testing
```
>>> int(Decimal("0.652814980819318111") * 10 ** 18)
652814980819318111
```
Ok, the answer is correct

```
>>> int(Decimal(amount) * 10 ** 18)
652814980819318102
```
Not correct, didn't convert to string

```
>>> int(Decimal(str(amount)) * 10 ** 18)
652814980819318100
```
Not correct either, but converted to string.
Can't believe its so difficult to do this with Python

**Comment by ErikBjare** [comment]
You can't assign a variable a simple float value with that much precision, the precision is dropped when it's assigned. That's basic float math, not Python specific.

```py
>>> amount = 0.652814980819318111
>>> amount
0.6528149808193181
```

Note how precision is already lost, this is why you should probably keep to the int representation whenever you can.

**Comment by alb2001** [comment]
Sorry, true. I automatically just stored it into a variable without thinking further that this is actually what breaks it into a float type by default.

Unfortunately I noticed that the self.tokenamount_Entry_var.get() method, stores its variable using a TK.DoubleVar() method, from Tkinter which effectively is taking out precision and rounding the number.

So I am not sure what to do here, except storing that as a Tk.StingVar() and then converting that string to a decimal using the Decimal class

**Comment by alb2001** [comment]
Yes, need to do some more tests, but apparently that worked, althought I don't see it's the most elegant way to solve it using a StringVar() where there should be a decimal

**Comment by liquid-8** [comment]
The framework you're using may provide functionality for different data types encapsulation. If so, the solution is trivial. Otherwise, I would consider using another framework.
